src/enzyme_ad/jax/Passes/LLVMToArith.cpp

```diff
@@ -0,0 +1,280 @@
+#define GEN_PASS_DEF_CONVERTLLVMTOARITHPASS
+#include "src/enzyme_ad/jax/Passes/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+
+#define PASS_NAME "convert-llvm-to-arith"
+
+//===----------------------------------------------------------------------===//
+// Pass Definition
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+template <typename SourceOp, typename TargetOp>
+class AttrConvertPassThrough {
+public:
+  AttrConvertPassThrough(SourceOp srcOp) : srcAttrs(srcOp->getAttrs()) {}
+
+  ArrayRef<NamedAttribute> getAttrs() const { return srcAttrs; }
+
+private:
+  ArrayRef<NamedAttribute> srcAttrs;
+};
+
+namespace {
+LogicalResult oneToOneRewrite(Operation *op, StringRef targetOp,
+                              ValueRange operands,
+                              ArrayRef<NamedAttribute> targetAttrs,
+                              PatternRewriter &rewriter) {
+  unsigned numResults = op->getNumResults();
+
+  // Create the operation through state since we don't know its C++ type.
+  Operation *newOp =
+      rewriter.create(op->getLoc(), rewriter.getStringAttr(targetOp), operands,
+                      op->getResultTypes(), targetAttrs);
+
+  // If the operation produced 0 or 1 result, return them immediately.
+  if (numResults == 0)
+    return rewriter.eraseOp(op), success();
+  if (numResults == 1)
+    return rewriter.replaceOp(op, newOp->getResult(0)), success();
+
+  return failure();
+}
+} // namespace
+
+template <typename TargetOp, typename SourceOp,
+          template <typename, typename> typename AttrConvert =
+              AttrConvertPassThrough>
+class VectorConvertToArithPattern : public OpRewritePattern<SourceOp> {
+public:
+  using OpRewritePattern<SourceOp>::OpRewritePattern;
+  using Super = VectorConvertToArithPattern<SourceOp, TargetOp>;
+
+  LogicalResult matchAndRewrite(SourceOp op,
+                                PatternRewriter &rewriter) const override {
+    static_assert(
+        std::is_base_of<OpTrait::OneResult<SourceOp>, SourceOp>::value,
+        "expected single result op");
+
+    SmallVector<Value> operands;
+    if constexpr (std::is_same<SourceOp, LLVM::ICmpOp>::value) {
+      for (auto opr : op->getOperands()) {
+        if (isa<LLVM::LLVMPointerType>(opr.getType()))
+          // TODO not always i64
+          operands.push_back(rewriter.create<LLVM::PtrToIntOp>(
+              op->getLoc(), rewriter.getI64Type(), opr));
+        else
+          operands.push_back(opr);
+      }
+    } else {
+      for (auto opr : op->getOperands())
+        operands.push_back(opr);
+    }
+
+    // Determine attributes for the target op
+    AttrConvert<SourceOp, TargetOp> attrConvert(op);
+
+    // TODO overflow flags
+    return oneToOneRewrite(op, TargetOp::getOperationName(), operands,
+                           attrConvert.getAttrs(), rewriter);
+  }
+};
+} // namespace
+
+#define AttrConvertOverflowToArith AttrConvertPassThrough
+
```

Comment:
minor, but for ease can we merge this with https://github.com/EnzymeAD/Enzyme-JAX/blob/b0663d35bf308b73b9593d9f0f5c3d732eb58ec3/src/enzyme_ad/jax/Passes/LibDeviceFuncsRaisingPass.cpp#L426 (that doesn't have cmp, but does have addf)